File: CSharp/TempurerClientCredentialsExample.cs
#define DEVELOPMENT_ONLY

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;
using System.Web;
using Newtonsoft.Json.Linq;  // Available from http://json.codeplex.com/ for parsing server responses.

namespace Tempurer.Web.Api.Example
{
    /// <summary>
    /// C# API integration sample v1.0
    /// </summary>
    public class TempurerClientCredentialsExample
    {
        // Set your client name and secret from http://www.tempurer.com/API/Applications here:
        private const string _clientName = "TODO";
        private const string _clientSecret = "TODO";

        // Production api and authorisation urls
        private const string _baseApiUrl = "api.tempurer.com";
        private const string _baseAuthUrl = "auth.tempurer.com";
        private readonly string _getAccessTokenUrl = String.Format("https://{0}/oauth2/token", _baseAuthUrl);

        public static void Main(string[] args)
        {
#if DEVELOPMENT_ONLY
            // Development only to avoid "Could not establish trust relationship for the SSL/TLS secure channel" when hosting authentication server in local IIS server with self signed https certificate.
            // Not to be used in production.
            ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
#endif
            var application = new TempurerClientCredentialsExample();
            try
            {
                application.ListVacancies();
                application.SearchVacancies();
                application.SummaryVacancies();
                application.GetVacancy();
            }
            catch (WebException ex)
            {
                Console.WriteLine(String.Format("HTTP Status: {0}", ex.Status));
                Console.WriteLine(String.Format("WebException thrown: {0}", ex));
            }
            catch (Exception ex)
            {
                Console.WriteLine(String.Format("Exception thrown: {0}", ex));
            }
        }

        /// <summary>
        /// Call to /services/vacancy.svc/list
        /// </summary>
        public void ListVacancies()
        {
            var scopes = new List<string>() { "basic", "/services/vacancy.svc/list" };
            var serviceUrl = String.Format("http://{0}/services/vacancy.svc/list", _baseApiUrl);

            var httpReturnValue = CallApiServiceUsingHttp(scopes, serviceUrl);
            
            Debug.WriteLine(httpReturnValue);
            Console.WriteLine(httpReturnValue);
        }

        /// <summary>
        /// Call to /services/vacancy.svc/search
        /// </summary>
        public void SearchVacancies()
        {
            var scopes = new List<string>() { "basic", "/services/vacancy.svc/search" };
            var serviceUrl = String.Format("http://{0}/services/vacancy.svc/search?query=test", _baseApiUrl);

            var httpReturnValue = CallApiServiceUsingHttp(scopes, serviceUrl);

            Debug.WriteLine(httpReturnValue);
            Console.WriteLine(httpReturnValue);
        }

        /// <summary>
        /// Call to /services/vacancy.svc/getsummary
        /// </summary>
        public void SummaryVacancies()
        {
            var scopes = new List<string>() { "basic", "/services/vacancy.svc/getsummary" };
            var serviceUrl = String.Format("http://{0}/services/vacancy.svc/getsummary", _baseApiUrl);

            var httpReturnValue = CallApiServiceUsingHttp(scopes, serviceUrl);

            Debug.WriteLine(httpReturnValue);
            Console.WriteLine(httpReturnValue);
        }

        /// <summary>
        /// Call to /services/vacancy.svc/get
        /// </summary>
        public void GetVacancy()
        {
            var scopes = new List<string>() { "basic", "/services/vacancy.svc/get" };
            var serviceUrl = String.Format("http://{0}/services/vacancy.svc/get/faf03b36-b42e-e211-8a1b-842b2b6577a5", _baseApiUrl);

            var httpReturnValue = CallApiServiceUsingHttp(scopes, serviceUrl);

            Debug.WriteLine(httpReturnValue);
            Console.WriteLine(httpReturnValue);
        }

        /// <summary>
        /// Call Api using pure HTTP calls only.
        /// </summary>
        private string CallApiServiceUsingHttp(IEnumerable<string> scopes, string serviceUrl)
        {
            Console.WriteLine(String.Format("\n\nCalling API URL: {0} with Scope: {1}", serviceUrl, String.Join(" ", serviceUrl)));

            // ------------------------------------------------
            // STEP 1: GetClientAccessToken
            // ------------------------------------------------

            // Create POST data and convert it to a byte array.
            Console.WriteLine("Auth token url: " + _getAccessTokenUrl);
            var request1 = WebRequest.Create(_getAccessTokenUrl);
            request1.Method = "POST";
            request1.ContentType = "application/x-www-form-urlencoded";

            var postDataString = "grant_type=client_credentials";
            var scopesString = HttpUtility.UrlEncode(String.Join(" ", scopes));
            if (scopesString.Length > 0)
            {
                postDataString = postDataString + "&scope=" + scopesString;
            }

            var postData = Encoding.UTF8.GetBytes(postDataString);
            request1.ContentLength = postData.Length;

            var base64details = Base64Encode(String.Format("{0}:{1}", _clientName, _clientSecret));
            var authHeader = String.Format("Basic {0}", base64details);
            request1.Headers[HttpRequestHeader.Authorization] = authHeader;

            // Get the request stream.
            var dataStream = request1.GetRequestStream();
            dataStream.Write(postData, 0, postData.Length);
            dataStream.Close();

            // Get the response.
            var response1 = request1.GetResponse();
            var status = ((HttpWebResponse)response1).StatusDescription;
            dataStream = response1.GetResponseStream();
            var reader1 = new StreamReader(dataStream);
            var responseFromServer1 = reader1.ReadToEnd();
            reader1.Close();
            if (dataStream != null)
            {
                dataStream.Close();
            }
            response1.Close();

            var json = JObject.Parse(responseFromServer1);   // Server response string: { "access_token":"your_token_value", "token_type":"bearer", "expires_in":"3600", "scope":"basic \/services\/vacancy.svc\/list" }
            var accessToken = json["access_token"].Value<string>();


            // ------------------------------------------------
            // STEP 2: Call webservice with Auth token in Header
            // ------------------------------------------------
            Console.WriteLine("Api token url: " + serviceUrl);
            var request2 = (HttpWebRequest)WebRequest.Create(serviceUrl);
            request2.Headers[HttpRequestHeader.Authorization] = String.Format("Bearer {0}", accessToken);

            var response2 = (HttpWebResponse)request2.GetResponse();
            var reader2 = new StreamReader(response2.GetResponseStream());
            var responseFromServer2 = reader2.ReadToEnd();
            return responseFromServer2;
        }

        private string Base64Encode(string input)
        {
            var toEncodeAsBytes = Encoding.ASCII.GetBytes(input);
            return Convert.ToBase64String(toEncodeAsBytes);
        }
    }
}

File: PHP/TempurerClientCredentialsExample.php
<?php 
/**
 *
 *	PHP API integration sample v1.0
 *
 *  The sample code below uses the Light PHP wrapper for the OAuth 2.0 library available from https://github.com/adoy/PHP-OAuth2
 *
 *	Steps:
 *  1) Get an api key and secret for your company by going to the application list at http://www.tempurer.com/API/Applications
 *	2) Set the  CLIENT_ID and CLIENT_SECRET values below
 *	
 *	Note that multiple scopes can be requested (separated by a space) in an Access Token request.  This allows the token to be re-used for different API calls.
 *
 */

	require('client.php');
	require('GrantType/IGrantType.php');
	require('GrantType/ClientCredentials.php');


	// Enter your API application credentials created at http://www.tempurer.com/API/Applications here:
	const CLIENT_ID     = 'TODO';
	const CLIENT_SECRET = 'TODO';
	

	// Constants to leave untouched
	const TOKEN_ENDPOINT       			= 'https://auth.tempurer.com/oauth2/token';
	const GRANT_TYPE 					= 'client_credentials';
	const AUTH_TYPE_AUTHORIZATION_BASIC = 1;
	const ACCESS_TOKEN_BEARER   		= 1;
	
	// Call api services
	ListVacancies();
	SearchVacancies();
	SummaryVacancies();
	GetVacancy();


	function ListVacancies()
	{
		$apiurl = 'http://api.tempurer.com/services/vacancy.svc/list';
		$scopes = 'basic /services/vacancy.svc/list';
		CallApiServiceUsingHttp($scopes, $apiurl);
	}


	function SearchVacancies()
	{
		$apiurl = 'http://api.tempurer.com/services/vacancy.svc/search?query=Test&page=1';
		$scopes = 'basic /services/vacancy.svc/search';
		CallApiServiceUsingHttp($scopes, $apiurl);
	}


	function SummaryVacancies()
	{
		$apiurl = 'http://api.tempurer.com/services/vacancy.svc/getsummary';
		$scopes = 'basic /services/vacancy.svc/getsummary';
		CallApiServiceUsingHttp($scopes, $apiurl);
	}


	function GetVacancy()
	{
		$apiurl = 'http://api.tempurer.com/services/vacancy.svc/get/faf03b36-b42e-e211-8a1b-842b2b6577a5';  // Set the guid of the vacancy you want to retrieve here.
		$scopes = 'basic /services/vacancy.svc/get';
		CallApiServiceUsingHttp($scopes, $apiurl);
	}


	// Generic function to call api after obtaining credentials from authorisation server.
	// Note that in a production system it would be more efficient to store the access token for reuse rather that obtaining it for every api call.
	function CallApiServiceUsingHttp($scopes, $apiurl)
	{
		// Create client object
		$client = new OAuth2\Client(CLIENT_ID, CLIENT_SECRET, AUTH_TYPE_AUTHORIZATION_BASIC);
		$client->setAccessTokenType(ACCESS_TOKEN_BEARER);

		// Get access token from authentication server
		$getAccessTokenParams = array('scope' => $scopes);
		$response = $client->getAccessToken(TOKEN_ENDPOINT, 'client_credentials', $getAccessTokenParams);
		$response_code = $response['code'];
		printf("%d", $response_code);
		
		printf("\n\nCALLING URL: %s WITH SCOPE: %s\n", $apiurl, $scopes);
		
		if ($response_code == 200) {
			// Success getting access token
			printf("SUCCESS GETTING TOKEN.\n");
			$access_token = $response['result']['access_token'];
			
			// Set access token on client object
			$client->setAccessToken($access_token);

			// Call api service
			$fetchParams = array('pagesize' => '1');
			$data = $client->fetch($apiurl, $fetchParams);
			var_dump($data);
		}
		else
		{
			// Error getting access token
			printf("ERROR OBTAINING TOKEN. CHECK CREDENTIALS, SCOPES AND TOKEN ENDPOINT.\n");
			//var_dump($response);
		}
	}

?> 